feat!: Enable access logging by default on Application Load Balancers (#2742)
guardian/cdk: .changeset/great-eyes-tan.md

---
"@guardian/cdk": major
---

> The `access_logs.s3.enabled` Application Load Balancer attribute is explicitly set, with a default value of `true`.

Author: I'd like to keep this as it indicates that, regardless of the value of `withAccessLogs`, a snapshot update is required. Agreed, it's too implementation detail-y for a subject line! Will move it down.

Access logging for Application Load Balancers (ALBs) is now enabled by default.
These logs are helpful during incident triage; therefore, they are now enabled by default.

> Previously, access logging was opt-in via the `accessLogging` object property on the `GuEc2App` pattern and its subclasses.

Reviewer: Minor point, but I think we could make life easier for users here and just name the relevant patterns (as knowing about the subclasses requires knowledge of this library code IIUC):

```suggestion
Previously users of the `GuEc2App`, `GuNodeApp`, `GuPlayApp ` and `GuEc2AppExperimental` patterns could opt-in to this logging via the `accessLogging` property.
```

Consequently, the optional `accessLogging` object property has been removed.
In its place is a new optional boolean property `withAccessLogging` which defaults to `true`.
- When `true` the ALB will have access logs enabled, configured to write to the account's S3 bucket
using a specific prefix for compatibility with the Athena database created in https://github.com/guardian/aws-account-setup.
- When `false` the `access_logs.s3.enabled` attribute is now explicitly set to `false`.

See also https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-properties-elasticloadbalancingv2-loadbalancer-loadbalancerattribute.html.

A `withAccessLogging` property is also added to the `GuApplicationLoadBalancer` construct.

NOTE: This feature requires a region to be set at the `GuStack` level, else the following error will be thrown:

> ValidationError: Region is required to enable ELBv2 access logging

> Here's an example of how to set the region:

Reviewer: 🎉

```typescript
class MyStack extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);
  }
}

const stackInstance = new MyStack(app, 'MyStack', {
  env: {
    region: 'eu-west-1'
  },
});
```

The main cost of this feature is incurred at query time (via Athena).

> There are no storage costs, as they are absorbed by AWS.

Reviewer: I think there is some storage cost associated with this (i.e. GBs stored in S3), even though AWS covers the cost of write requests from the ELB service?

If that's correct, we might want to mention it just in case it has an impact on very high traffic services like Ophan? But we could definitely mention that for most services the storage costs are negligible?

Author: Ah, good point! From https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-access-logs.html:

> You are charged storage costs for Amazon S3, but not charged for the bandwidth used by Elastic Load Balancing to send log files to Amazon S3. For more information about storage costs

---
"@guardian/cdk": major
---

Access logging for Application Load Balancers (ALBs) is now enabled by default.

The Application Load Balancer (ALB) access logs are helpful during incident triage.
Therefore, they are now enabled by default.

Previously users of the `GuEc2App`, `GuNodeApp`, `GuPlayApp `, `GuPlayWorkerApp` and `GuEc2AppExperimental` patterns could opt-in to this logging via the `accessLogging` property and configure the S3 prefix.

This property is now removed and replaced with a new optional boolean property `withAccessLogging` which defaults to `true`.
- When `true` the ALB will have access logs enabled, configured to write to the account's S3 bucket

> using a specific prefix for compatibility with the Athena database created in https://github.com/guardian/aws-account-setup.

Reviewer 2: Maybe this isn't the right place to mention it, but I think it'd be useful to write the name of the database somewhere.

```suggestion
using a specific prefix for compatibility with the `gucdk_access_logs` database created in Athena via https://github.com/guardian/aws-account-setup.
```

Author: Updated. I've also added this to the docstrings too.
